### PR TITLE
feat: `LinearMap.submoduleMap`

### PR DESCRIPTION
**Pull request by llllvvuu**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Submodule/Map.lean [hunk 1]
```diff
@@ -619,3 +619,25 @@ def compatibleMaps : Submodule R (N →ₗ[R] N₂) where
...
+
+variable { M M' R : Type* } [Semiring R] [AddCommMonoid M] [AddCommMonoid M']
+  [Module R M] [Module R M']
+
+/-- A linear map between two modules restricts to a linear map from any submodule p of the
+domain onto the image of that submodule.
+
+This is the linear version of AddMonoidHom.addSubmonoidMap and AddMonoidHom.addSubgroupMap.-/
+def submoduleMap (f : M →ₗ[R] M') (S : Submodule R M) : S →ₗ[R] S.map f where
+  toFun := fun x ↦ ⟨ f x, Submodule.apply_coe_mem_map f x ⟩
```
Comment (alreadydone) [inline Mathlib/Algebra/Module/Submodule/Map.lean:633]: Note that [LinearEquiv.submoduleMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Basic.html#LinearEquiv.submoduleMap) is defined using domRestrict and codRestrict.

File: Mathlib/Algebra/Module/Submodule/Map.lean [hunk 2]
```diff
@@ -619,3 +619,25 @@ def compatibleMaps : Submodule R (N →ₗ[R] N₂) where
...
+
+namespace LinearMap
+
+variable { M M' R : Type* } [Semiring R] [AddCommMonoid M] [AddCommMonoid M']
+  [Module R M] [Module R M']
+
+/-- A linear map between two modules restricts to a linear map from any submodule p of the
+domain onto the image of that submodule.
+
+This is the linear version of AddMonoidHom.addSubmonoidMap and AddMonoidHom.addSubgroupMap.-/
```
Comment (eric-wieser) [inline Mathlib/Algebra/Module/Submodule/Map.lean:631]: Thanks for these cross-references, they're very useful to have in docstrings.

If you put them in backticks, docgen will link them:
```suggestion
This is the linear version of `AddMonoidHom.addSubmonoidMap` and `AddMonoidHom.addSubgroupMap`. -/
```

File: Mathlib/Algebra/Module/Submodule/Map.lean [hunk 3]
```diff
@@ -619,3 +619,25 @@ def compatibleMaps : Submodule R (N →ₗ[R] N₂) where
 #align submodule.compatible_maps Submodule.compatibleMaps
 
 end Submodule
+
+namespace LinearMap
+
+variable { M M' R : Type* } [Semiring R] [AddCommMonoid M] [AddCommMonoid M']
+  [Module R M] [Module R M']
+
+/-- A linear map between two modules restricts to a linear map from any submodule p of the
```
Comment (eric-wieser) [inline Mathlib/Algebra/Module/Submodule/Map.lean:628]: Your function argument is called `S`:
```suggestion
/-- A linear map between two modules restricts to a linear map from any submodule `S` of the
```
(or change the function argument instead)

File: Mathlib/Algebra/Module/Submodule/Map.lean [hunk 4]
```diff
@@ -619,3 +619,25 @@ def compatibleMaps : Submodule R (N →ₗ[R] N₂) where
 #align submodule.compatible_maps Submodule.compatibleMaps
 
 end Submodule
+
+namespace LinearMap
+
+variable { M M' R : Type* } [Semiring R] [AddCommMonoid M] [AddCommMonoid M']
+  [Module R M] [Module R M']
```
Comment (eric-wieser) [inline Mathlib/Algebra/Module/Submodule/Map.lean:626]: We already have
```
`variable {R : Type*} {R₁ : Type*} {R₂ : Type*} {R₃ : Type*}
variable {S : Type*}
variable {M : Type*} {M₁ : Type*} {M₂ : Type*} {M₃ : Type*}
```
at the top of the file; please reuse these.
```suggestion
variable [Semiring R] [AddCommMonoid M] [AddCommMonoid M₁] [Module R M] [Module R M₁]
```
or similar.

**Comment by llllvvuu** [comment]
Thanks @eric-wieser @alreadydone! updated

**Comment by llllvvuu** [comment]
Closing for #8963 since this is from a fork

**Pull request closed without merging**